source/solver_class.cc

```diff
@@ -83,6 +213,757 @@ void ProjectionSolverBase<dim>::set_supply_term(
+    (lagrange_multiplier->get_locally_owned_dofs(),
+     lagrange_multiplier->get_locally_owned_dofs(),
+     lagrange_multiplier_sparsity_pattern,
+     this->mpi_communicator);
+    projection_step_system_matrix.reinit
+    (auxiliary_scalar->get_locally_owned_dofs(),
+     auxiliary_scalar->get_locally_owned_dofs(),
+     auxiliary_scalar_sparsity_pattern,
+     this->mpi_communicator);
+
+  #else
+    TrilinosWrappers::SparsityPattern
+    lagrange_multiplier_sparsity_pattern(
+      lagrange_multiplier->get_locally_owned_dofs(),
+      lagrange_multiplier->get_locally_owned_dofs(),
+      lagrange_multiplier->get_locally_relevant_dofs(),
+      this->mpi_communicator);
+
+    TrilinosWrappers::SparsityPattern
+    auxiliary_scalar_sparsity_pattern(
+      auxiliary_scalar->get_locally_owned_dofs(),
+      auxiliary_scalar->get_locally_owned_dofs(),
+      auxiliary_scalar->get_locally_relevant_dofs(),
+      this->mpi_communicator);
+
+    DoFTools::make_sparsity_pattern(
+      lagrange_multiplier->get_dof_handler(),
+      lagrange_multiplier_sparsity_pattern,
+      lagrange_multiplier->get_constraints(),
+      false,
+      Utilities::MPI::this_mpi_process(this->mpi_communicator));
+
+    DoFTools::make_sparsity_pattern(
+      auxiliary_scalar->get_dof_handler(),
+      auxiliary_scalar_sparsity_pattern,
+      auxiliary_scalar->get_constraints(),
+      false,
+      Utilities::MPI::this_mpi_process(this->mpi_communicator));
+
+    lagrange_multiplier_sparsity_pattern.compress();
+    auxiliary_scalar_sparsity_pattern.compress();
+
+    zeroth_step_system_matrix.reinit(lagrange_multiplier_sparsity_pattern);
+    projection_step_system_matrix.reinit(auxiliary_scalar_sparsity_pattern);
+
+  #endif
+}
+
+
+
+template <int dim>
+void ProjectionSolverBase<dim>::setup_vectors()
+{
+  diffusion_step_rhs.reinit(vector_field->distributed_vector);
+  projection_step_rhs.reinit(auxiliary_scalar->distributed_vector);
+  zeroth_step_rhs.reinit(lagrange_multiplier->distributed_vector);
+}
+
+
+
+template <int dim>
+void ProjectionSolverBase<dim>::zeroth_step()
+{
+  if (ptr_supply_term != nullptr)
+    ptr_supply_term->set_time(this->time_stepping.get_start_time());
+
+  assemble_zeroth_step();
+
+  solve_zeroth_step();
+}
+
+
+
+template <int dim>
+std::pair<int, double> ProjectionSolverBase<dim>::solve_zeroth_step()
+{
+  LinearAlgebra::MPI::Vector distributed_lagrange_multiplier(lagrange_multiplier->distributed_vector);
+  distributed_lagrange_multiplier = lagrange_multiplier->old_solution;
+
+  const typename RunTimeParameters::LinearSolverParameters &solver_parameters
+    = projection_solver_parameters.zeroth_step_solver_parameters;
+
+  build_preconditioner(zeroth_step_preconditioner,
+                       zeroth_step_system_matrix,
+                       solver_parameters.preconditioner_parameters_ptr,
+                       (lagrange_multiplier->fe_degree() > 1? true: false));
+
+  AssertThrow(zeroth_step_preconditioner != nullptr,
```

Comment:
Or equivalent `AssertThrow(zeroth_step_preconditioner, ...)` for a smart pointer.